Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1297 @@
+
+```happy
+exp :: { Exp L.Range }
+  : exp atom                 { EApp (info $1 <-> info $2) $1 $2 }
+  | atom                     { $1 }
+
+atom :: { Exp L.Range }
+  : integer                  { unTok $1 (\range (L.Integer int) -> EInt range int) }
+  | name                     { EVar (info $1) $1 }
+  | string                   { unTok $1 (\range (L.String string) -> EString range string) }
+  | '(' ')'                  { EUnit (L.rtRange $1 <-> L.rtRange $2) }
+  | '[' sepBy(exp, ',') ']'  { EList (L.rtRange $1 <-> L.rtRange $3) $2 }
+  | '(' exp ')'              { EPar (L.rtRange $1 <-> L.rtRange $3) $2 }
+```
+
+This will cause Happy to build a chain `(((atom atom) atom) atom) atom` when parsing.
+Note that even though the left side was stated to be an `exp`, eventually, we must reach an `atom` to terminate parsing.
+
+With it, function applications should now be left-balanced.
+
+#### Parsing conditional expressions
+
+Let's see another case where we will get ambiguities.
+Adding these rules to `exp` will indicate 11 shift/reduce conflicts:
+
+```happy
+  | if exp then exp          { EIfThen (L.rtRange $1 <-> info $4) $2 $4 }
+  | if exp then exp else exp { EIfThenElse (L.rtRange $1 <-> info $6) $2 $4 $6 }
+```
+
+If you see the `Parser.info` file, most ambiguities come from two states. The first one is this:
+
+```text
+State 51
+
+    exp -> exp . atom                                   (rule 12)
+    exp -> if exp then exp .                            (rule 13)
+    exp -> if exp then exp . else exp                   (rule 14)
+```
+
+We see that there is an ambiguity when parsing function applications and conditionals.
+One instance of a conflicting parse is `if b then f x` when we arrive at `if b then f . x`.
+Should it be parsed as `if b then (f x)` or `(if b then f) x`?
+
+Shifting will cause the first parse to be accepted while reducing will cause the second parse to be accepted.
+Remember, intuitively, shifting can be approximated as "continue parsing and see what we get" while reducing as "accept what we have so far".
+This is why reducing inserts parenthesis right after `f` is parsed.
+
+For now, just like we've split `exp` into `exp` and `atom` in the past, we can further split it:
+
+```happy
+exp :: { Exp L.Range }
+  : expapp                   { $1 }
+  | expcond                  { $1 }
+
+expapp :: { Exp L.Range }
+  : expapp atom              { EApp (info $1 <-> info $2) $1 $2 }
+  | atom                     { $1 }
+
+expcond :: { Exp L.Range }
+  : if exp then exp          { EIfThen (L.rtRange $1 <-> info $4) $2 $4 }
+  | if exp then exp else exp { EIfThenElse (L.rtRange $1 <-> info $6) $2 $4 $6 }
+```
+
+We use `expapp` to represent function applications, which will create a chain of `atom`s that represents the application.
+Likewise, `expcond` represents the conditional expressions.
+Since the left-hand side of an application can't be an `if` expression anymore, the conflict is eliminated.
```

Comment:
That is not true. You can just surround `if` expression with parentheses

Reply:
You're right, that was a poor choice of words on my part. I rephrased it.